project2-beat-that/script.js

```diff
@@ -1,4 +1,299 @@
+    if (input == `SCORE`){
+      gameState = SCORE;
+      console.log(`Game state = ${gameState}`);
+    }
+    // else if (input == `LOWEST`){
+    //   gameState = LOWEST_COMBINE_NUMBDER;
+    //   console.log(`Game state = ${gameState}`);
+    
+    // }
+    else {
+      gameState = ROLL_DICE; 
+    }
+    return introMessage(); 
+  }
+  
+  // PLAYER 1 STARTS
+  if (gameState == ROLL_DICE){
+    if (input == `SCORE`){
+      gameState = SCORE;
+      console.log(`Game state = ${gameState}`);
+    }
+    // else if (input == `LOWEST`){
+    //   gameState = LOWEST_COMBINE_NUMBDER;
+    //   console.log(`Game state = ${gameState}`);
+    // }
+    else{
+      gameState = CHOOSE_DICE_ORDER;
+    }
+    
+    
+    return playerOneRollDice(); 
+  }
+  
+  // Player 1 chooses dice order 
+  if (gameState == CHOOSE_DICE_ORDER){
+    console.log(`New game state: ${gameState}`);
+    if (input == '1'){
+      playerOneNumber = String(playerOne[0]) + String(playerOne[1]);
+      playerOneCombine.push(playerOneNumber);
+      gameState = CHANGE_TO_PLAYER_2;
+      output = `Player 1 number is ${playerOneNumber}. <br> 
+      Player 2 please press submit to generate your numbers `
+      return output;
+    }
+    else if (input == '2'){
+      playerOneNumber = String(playerOne[1]) + String(playerOne[0]);
+      playerOneCombine.push(playerOneNumber);
+      gameState = CHANGE_TO_PLAYER_2;
+      output = `Player 1 number is ${playerOneNumber}. <br> 
+      Player 2 please press submit to generate your numbers `
+      return output;
+    }
+    else if (input == `SCORE`){
+      gameState = SCORE;
+    }
+    else if (input == `LOWEST`){
+      gameState = LOWEST_COMBINE_NUMBDER;
+    }
+    else{
+      return `Invalid response, key in "1" or "2"`;
+    }
+  }
+
+  // PLAYER 2 STARTS 
+  if (gameState == CHANGE_TO_PLAYER_2){
+    console.log(`Player 2 mode`);
+    gameState = CHOOSE_DICE_ORDER_PLAYER_2;
+    return playerTwoRollDice();
+  }
+
+// player 2 selects order 
+  if (gameState == CHOOSE_DICE_ORDER_PLAYER_2){
+    console.log(`New game state: ${gameState}`);
+    if (input == '1'){
+      gameState = COMPARE_PLAYERS;
+      playerTwoNumber = String(playerTwo[0]) + String(playerTwo[1]);
+      playerTwoCombine.push(playerTwoNumber);
+      output = `Player 1 number is ${playerOneNumber}. <br>
+      Player 2 number is ${playerTwoNumber}. <br>
+      Press submit to find out who wins!`
+      return output;
+    }
+    else if (input == '2'){
+      gameState = COMPARE_PLAYERS;
+      playerTwoNumber = String(playerTwo[1]) + String(playerTwo[0]);
+      playerTwoCombine.push(playerTwoNumber);
+      output = `Player 1 number is ${playerOneNumber}. <br>
+      Player 2 number is ${playerTwoNumber}. <br>
+      Press submit to find out who wins! `
+      return output;
+    }
+    else if (input == `SCORE`){
+      gameState = SCORE;
+    }
+   
+    else{
+      return `Invalid response, key in "1" or "2"`;
+    }
+  }
```

Comment:
also good but same comments about repeated logic as above (in lines 150-167) and on shifting the input validation up

Reply:
again, this is to ensure scalability of the code and also to reduce repetition. as programmers we try to repeat logic as little as possible to make the code succinct. but also imagine if you have a bug somewhere in this logic, you need to change it in 2 places, and possibly even more if there are more instances of repeated logic in your code